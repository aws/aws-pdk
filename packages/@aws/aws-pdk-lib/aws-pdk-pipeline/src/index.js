"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
__exportStar(require("./pdk-pipeline"), exports);
__exportStar(require("./pdk-pipeline-ts-project"), exports);
__exportStar(require("./pdk-pipeline-py-project"), exports);
__exportStar(require("./pdk-pipeline-java-project"), exports);
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyJpbmRleC50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOzs7Ozs7Ozs7Ozs7QUFBQSxpREFBK0I7QUFDL0IsNERBQTBDO0FBQzFDLDREQUEwQztBQUMxQyw4REFBNEMiLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgKiBmcm9tIFwiLi9wZGstcGlwZWxpbmVcIjtcbmV4cG9ydCAqIGZyb20gXCIuL3Bkay1waXBlbGluZS10cy1wcm9qZWN0XCI7XG5leHBvcnQgKiBmcm9tIFwiLi9wZGstcGlwZWxpbmUtcHktcHJvamVjdFwiO1xuZXhwb3J0ICogZnJvbSBcIi4vcGRrLXBpcGVsaW5lLWphdmEtcHJvamVjdFwiO1xuIl19