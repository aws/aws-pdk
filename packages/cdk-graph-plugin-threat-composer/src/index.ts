/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./plugin";
export * from "./model-generator/types";
