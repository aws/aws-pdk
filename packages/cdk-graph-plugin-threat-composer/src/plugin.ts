/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as fs from "fs";
import {
  CdkGraph,
  CdkGraphArtifact,
  CdkGraphContext,
  ICdkGraphPlugin,
  IGraphPluginBindCallback,
  IGraphReportCallback,
} from "@aws/cdk-graph";
import { PDKNagApp } from "@aws/pdk-nag";
import { ThreatModelGenerator } from "./model-generator/threat-model-generator";
import { ThreatComposerApplicationDetails } from "./model-generator/types";

/**
 * Options for the Threat Composer CDK Graph plugin
 */
export interface CdkGraphThreatComposerPluginOptions {
  /**
   * Details about the application to include in the threat model
   */
  readonly applicationDetails?: ThreatComposerApplicationDetails;
}

/**
 * CdkGraphThreatComposerPlugin is a {@link ICdkGraphPlugin CdkGraph Plugin} implementation for generating
 * Threat Composer threat models.
 * @see https://github.com/awslabs/threat-composer
 */
export class CdkGraphThreatComposerPlugin implements ICdkGraphPlugin {
  /**
   * Fixed ID of the threat-composer plugin
   */
  static readonly ID = "threat-composer";

  /**
   * Curent semantic version of the threat-composer plugin
   */
  static readonly VERSION = "0.0.0";

  /** @inheritdoc */
  get id(): string {
    return CdkGraphThreatComposerPlugin.ID;
  }

  /** @inheritdoc */
  get version(): string {
    return CdkGraphThreatComposerPlugin.VERSION;
  }

  /**
   * Retrieve the threat model artifact
   */
  get threatModelArtifact(): CdkGraphArtifact | undefined {
    return this._threatModelArtifact;
  }

  // TODO: consider graph plugin as dependency?
  /** @inheritdoc */
  readonly dependencies?: string[] = [];

  /** @internal */
  private _app?: PDKNagApp;

  /** @internal */
  private _options?: CdkGraphThreatComposerPluginOptions;

  /** @internal */
  private _threatModelArtifact: CdkGraphArtifact | undefined = undefined;

  constructor(options?: CdkGraphThreatComposerPluginOptions) {
    this._options = options;
  }

  /** @inheritdoc */
  bind: IGraphPluginBindCallback = (graph: CdkGraph): void => {
    // Validate the top level node is a PDKNagApp, providing us with access to nag results
    if (!("extendedNagResults" in graph.root)) {
      throw new Error(
        `Threat Composer plugin requires the root CDK construct to be a PDKNagApp`
      );
    }
    this._app = graph.root as PDKNagApp;
  };

  /** @inheritdoc */
  report?: IGraphReportCallback = async (
    context: CdkGraphContext
  ): Promise<void> => {
    if (!this._app) {
      throw new Error("Plugin has not been bound");
    }

    let architectureImageDataUri: string | undefined = undefined;

    const architectureDiagramArtifact: CdkGraphArtifact | undefined =
      context.artifacts.DIAGRAM_PNG;
    if (architectureDiagramArtifact) {
      const diagramBinaryContent = fs.readFileSync(
        architectureDiagramArtifact.filepath
      );
      architectureImageDataUri = `data:image/png;base64,${diagramBinaryContent.toString(
        "base64"
      )}`;
    }

    const threatModel = new ThreatModelGenerator().generate(
      this._app.extendedNagResults(),
      {
        ...this._options,
        architectureImageDataUri,
      }
    );

    this._threatModelArtifact = context.writeArtifact(
      this,
      "THREAT_MODEL",
      "threat-model.tc.json",
      JSON.stringify(threatModel, null, 2),
      "Threat Composer threat model."
    );
  };
}
