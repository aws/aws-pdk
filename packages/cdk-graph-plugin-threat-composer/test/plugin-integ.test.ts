/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as fs from "fs";
import { CdkGraph } from "@aws/cdk-graph";
import { PDKPipelineIntegApp } from "@aws/cdk-graph/test/__fixtures__/pdk-integ";
import * as testUtils from "./test-utils";
import { CdkGraphThreatComposerPlugin } from "../src";

jest.setTimeout(90000);

const makeCdkOutdir = async (name: string) =>
  testUtils.makeCdkOutDir("plugin-integ", name);

describe("plugin-integ", () => {
  describe("pipeline", () => {
    let plugin: CdkGraphThreatComposerPlugin;

    beforeAll(async () => {
      const outdir = await makeCdkOutdir("pipeline");

      const app = new PDKPipelineIntegApp({ outdir });
      plugin = new CdkGraphThreatComposerPlugin();
      const graph = new CdkGraph(app, {
        plugins: [plugin],
      });
      app.synth();
      await graph.report();
    });

    it("should generate a threat model", () => {
      expect(plugin.threatModelArtifact).toBeDefined();
      expect(fs.existsSync(plugin.threatModelArtifact!.filepath)).toBeTruthy();
      expect(
        fs.readFileSync(plugin.threatModelArtifact!.filepath, "utf-8")
      ).toMatchSnapshot();
    });
  });
});
