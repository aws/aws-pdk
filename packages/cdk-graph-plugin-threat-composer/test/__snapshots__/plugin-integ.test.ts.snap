// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-integ pipeline should generate a threat model 1`] = `
"{
  "schema": 1,
  "applicationInfo": {
    "name": "My Application",
    "description": "**Important**: This threat model is not comprehensive or complete. This document does not list all possible threats, and where threats do have mapped mitigations - these mitigations are not deemed to be complete.\\n\\nConsider this as a starting point for your own diligence and threat modeling."
  },
  "architecture": {
    "image": "",
    "description": ""
  },
  "dataflow": {
    "image": "",
    "description": "**Note:** A data flow diagram has not been captured at this stage. You should consider creating a data-flow diagram."
  },
  "assumptions": [],
  "mitigations": [
    {
      "id": "89a72e46-d073-4207-bd5a-23636c9a29e6",
      "numericId": 1,
      "displayOrder": 1,
      "tags": [
        "CloudFront"
      ],
      "content": "cdk-nag rule: CloudFrontDistributionNoOutdatedSSL",
      "metadata": [
        {
          "key": "Comments",
          "value": "2 of 2 Resources Compliant."
        }
      ]
    },
    {
      "id": "41220750-87bd-4c85-94f1-999746809166",
      "numericId": 2,
      "displayOrder": 2,
      "tags": [
        "CloudFront"
      ],
      "content": "cdk-nag rule: CloudFrontDistributionS3OriginAccessIdentity",
      "metadata": [
        {
          "key": "Comments",
          "value": "2 of 2 Resources Compliant."
        }
      ]
    },
    {
      "id": "64a10a2f-8f11-4e98-8d08-8675853b9fef",
      "numericId": 3,
      "displayOrder": 3,
      "tags": [
        "CodeBuild"
      ],
      "content": "cdk-nag rule: CodeBuildProjectKMSEncryptedArtifacts",
      "metadata": [
        {
          "key": "Comments",
          "value": "3 of 3 Resources Compliant."
        }
      ]
    },
    {
      "id": "ea70a6ca-3fa7-403b-87fc-80e4da6bce1e",
      "numericId": 4,
      "displayOrder": 4,
      "tags": [
        "CodeBuild"
      ],
      "content": "cdk-nag rule: CodeBuildProjectManagedImages",
      "metadata": [
        {
          "key": "Comments",
          "value": "3 of 3 Resources Compliant."
        }
      ]
    },
    {
      "id": "c84d3721-1539-444a-af67-bd2fed80735f",
      "numericId": 5,
      "displayOrder": 5,
      "tags": [
        "CodeBuild"
      ],
      "content": "cdk-nag rule: CodeBuildProjectPrivilegedModeDisabled",
      "metadata": [
        {
          "key": "Comments",
          "value": "3 of 3 Resources Compliant."
        }
      ]
    },
    {
      "id": "0edef48e-478a-4354-afba-04cbd8a678af",
      "numericId": 6,
      "displayOrder": 6,
      "tags": [
        "EC2",
        "VPC"
      ],
      "content": "cdk-nag rule: EC2RestrictedInbound",
      "metadata": [
        {
          "key": "Comments",
          "value": "4 of 6 Resources Compliant."
        }
      ]
    },
    {
      "id": "a3adc4bc-12e6-4e51-9a3b-f70173626aed",
      "numericId": 7,
      "displayOrder": 7,
      "tags": [
        "EC2",
        "VPC"
      ],
      "content": "cdk-nag rule: EC2SecurityGroupDescription",
      "metadata": [
        {
          "key": "Comments",
          "value": "6 of 6 Resources Compliant."
        }
      ]
    },
    {
      "id": "da85ab97-eb06-4c15-9782-5a7a1924c7d4",
      "numericId": 8,
      "displayOrder": 8,
      "tags": [
        "IAM"
      ],
      "content": "cdk-nag rule: IAMNoManagedPolicies",
      "metadata": [
        {
          "key": "Comments",
          "value": "15 of 33 Resources Compliant."
        }
      ]
    },
    {
      "id": "ca417660-79a6-4725-b498-7406ffc8a9f5",
      "numericId": 9,
      "displayOrder": 9,
      "tags": [
        "IAM"
      ],
      "content": "cdk-nag rule: IAMNoWildcardPermissions",
      "metadata": [
        {
          "key": "Comments",
          "value": "29 of 48 Resources Compliant.\\n\\n__Suppression Reasons__:\\n* Wildcards are needed for dynamically created resources."
        }
      ]
    },
    {
      "id": "9cd8dfd8-8f1f-4697-aa2d-17378c3eb587",
      "numericId": 10,
      "displayOrder": 10,
      "tags": [
        "KMS"
      ],
      "content": "cdk-nag rule: KMSBackingKeyRotationEnabled",
      "metadata": [
        {
          "key": "Comments",
          "value": "1 of 1 Resources Compliant."
        }
      ]
    },
    {
      "id": "fcd9742a-30eb-4303-97e7-206e4a1f63fd",
      "numericId": 11,
      "displayOrder": 11,
      "tags": [
        "Lambda"
      ],
      "content": "cdk-nag rule: LambdaLatestVersion",
      "metadata": [
        {
          "key": "Comments",
          "value": "8 of 14 Resources Compliant."
        }
      ]
    },
    {
      "id": "0ae5ca12-88c7-4648-a1e8-28ad585c3336",
      "numericId": 12,
      "displayOrder": 12,
      "tags": [
        "S3"
      ],
      "content": "cdk-nag rule: S3BucketLevelPublicAccessProhibited",
      "metadata": [
        {
          "key": "Comments",
          "value": "6 of 6 Resources Compliant."
        }
      ]
    },
    {
      "id": "646feb9c-6f18-4307-8a90-8a44b851d176",
      "numericId": 13,
      "displayOrder": 13,
      "tags": [
        "S3"
      ],
      "content": "cdk-nag rule: S3BucketLoggingEnabled",
      "metadata": [
        {
          "key": "Comments",
          "value": "2 of 6 Resources Compliant."
        }
      ]
    },
    {
      "id": "b19fd780-64a3-42c6-a0d2-92833af17e40",
      "numericId": 14,
      "displayOrder": 14,
      "tags": [
        "S3"
      ],
      "content": "cdk-nag rule: S3BucketSSLRequestsOnly",
      "metadata": [
        {
          "key": "Comments",
          "value": "4 of 8 Resources Compliant."
        }
      ]
    },
    {
      "id": "c0f8d761-b616-49d4-b60e-d9d413c91f19",
      "numericId": 15,
      "displayOrder": 15,
      "tags": [
        "S3",
        "CloudFront"
      ],
      "content": "cdk-nag rule: S3WebBucketOAIAccess",
      "metadata": [
        {
          "key": "Comments",
          "value": "6 of 6 Resources Compliant."
        }
      ]
    }
  ],
  "assumptionLinks": [],
  "mitigationLinks": [
    {
      "mitigationId": "89a72e46-d073-4207-bd5a-23636c9a29e6",
      "linkedId": "c7d756f0-0576-4644-a9b3-afea5bd52a7c"
    },
    {
      "mitigationId": "41220750-87bd-4c85-94f1-999746809166",
      "linkedId": "d4378c9b-e680-41c1-8cc2-136e124461dd"
    },
    {
      "mitigationId": "64a10a2f-8f11-4e98-8d08-8675853b9fef",
      "linkedId": "7b1dbdd4-2cea-4b16-a15b-51b05822f972"
    },
    {
      "mitigationId": "ea70a6ca-3fa7-403b-87fc-80e4da6bce1e",
      "linkedId": "3e3ee2f1-053d-4e40-8917-52ce4f45ea56"
    },
    {
      "mitigationId": "c84d3721-1539-444a-af67-bd2fed80735f",
      "linkedId": "9d2c9e3b-f5c8-42af-8f63-4bd2309cf0d9"
    },
    {
      "mitigationId": "0edef48e-478a-4354-afba-04cbd8a678af",
      "linkedId": "8c3352e2-b817-4139-9256-81e75ff127f7"
    },
    {
      "mitigationId": "a3adc4bc-12e6-4e51-9a3b-f70173626aed",
      "linkedId": "310b7a7f-e13c-44a3-bb3a-91ad7bf9ba98"
    },
    {
      "mitigationId": "da85ab97-eb06-4c15-9782-5a7a1924c7d4",
      "linkedId": "1c62b52e-f3ba-4968-a14d-745691b706f6"
    },
    {
      "mitigationId": "ca417660-79a6-4725-b498-7406ffc8a9f5",
      "linkedId": "1c62b52e-f3ba-4968-a14d-745691b706f6"
    },
    {
      "mitigationId": "fcd9742a-30eb-4303-97e7-206e4a1f63fd",
      "linkedId": "5d521ccc-ca9c-4790-b045-0a1e26de9aa0"
    },
    {
      "mitigationId": "0ae5ca12-88c7-4648-a1e8-28ad585c3336",
      "linkedId": "8c383856-988d-45a4-bc92-5c9aab42aab3"
    },
    {
      "mitigationId": "646feb9c-6f18-4307-8a90-8a44b851d176",
      "linkedId": "1b240cb7-3557-41e7-9c19-21ebc73ad0ec"
    },
    {
      "mitigationId": "b19fd780-64a3-42c6-a0d2-92833af17e40",
      "linkedId": "77e519f0-641e-4b99-96a3-db0130381c97"
    },
    {
      "mitigationId": "c0f8d761-b616-49d4-b60e-d9d413c91f19",
      "linkedId": "cc8b9fcb-dd2a-41ad-a4f1-11e93ae0b031"
    }
  ],
  "threats": [
    {
      "id": "dc429030-4063-43db-96ea-d60b8ab83152",
      "numericId": 1,
      "displayOrder": 1,
      "metadata": [],
      "tags": [
        "API Gateway"
      ],
      "threatSource": "external threat actor",
      "prerequisites": "with a network path to the API",
      "threatAction": "make arbitrary web requests that are either not explicitly allowed, blocked or counted",
      "threatImpact": "them perform reconnaissance and intrusion activities against the API",
      "impactedAssets": [
        "this application"
      ],
      "statement": "An external threat actor with a network path to the API can make arbitrary web requests that are either not explicitly allowed, blocked or counted, which leads to them perform reconnaissance and intrusion activities against the API, negatively impacting this application"
    },
    {
      "id": "2d370353-bc79-43a7-8e92-68d0850a13d0",
      "numericId": 2,
      "displayOrder": 2,
      "metadata": [],
      "tags": [
        "API Gateway"
      ],
      "threatSource": "external threat actor",
      "prerequisites": "with a network path to the API",
      "threatAction": "make unauthenticated calls to the API",
      "threatImpact": "",
      "impactedGoal": [
        "confidentiality",
        "integrity",
        "availability"
      ],
      "impactedAssets": [
        "this application"
      ],
      "statement": "An external threat actor with a network path to the API can make unauthenticated calls to the API, resulting in reduced confidentiality, integrity and/or availability of this application"
    },
    {
      "id": "813c3fc1-33bc-4823-9f9c-91b2df1f4030",
      "numericId": 3,
      "displayOrder": 3,
      "metadata": [],
      "tags": [
        "API Gateway"
      ],
      "threatSource": "external threat actor",
      "prerequisites": "with a network path to, and authorization to use the API",
      "threatAction": "can supply unexpected HTTP headers or JSON payload to the API",
      "threatImpact": "the application behaving in unintended ways",
      "impactedGoal": [],
      "impactedAssets": [
        "this application"
      ],
      "statement": "An external threat actor with a network path to, and authorization to use the API can can supply unexpected HTTP headers or JSON payload to the API, which leads to the application behaving in unintended ways, negatively impacting this application"
    },
    {
      "id": "8c9cd385-7a3d-486d-90fc-5b9f3b28a04a",
      "numericId": 4,
      "displayOrder": 4,
      "tags": [
        "CloudFront"
      ],
      "threatSource": "external threat actor",
      "threatAction": "discover application web assets that are exposed via CloudFront distributions",
      "threatImpact": "reconnaissance and intrusion activities being performed against the exposed attack surface",
      "impactedAssets": [
        "this application"
      ],
      "statement": "An external threat actor can discover application web assets that are exposed via CloudFront distributions, which leads to reconnaissance and intrusion activities being performed against the exposed attack surface, negatively impacting this application"
    },
    {
      "id": "8cde9550-4aac-4792-851f-e6ac01b3ae92",
      "numericId": 5,
      "displayOrder": 5,
      "metadata": [
        {
          "key": "STRIDE",
          "value": [
            "T",
            "I"
          ]
        }
      ],
      "tags": [
        "CloudFront"
      ],
      "threatSource": "threat actor",
      "prerequisites": "who is in a person-in-the-middle position between the user and the CloudFront Distribution",
      "threatAction": "take advantage of a vulnerability in an older/depreciated SSL/TLS protocol",
      "threatImpact": "them being able manipulate view or modify the content served by the CloudFront Distribution",
      "impactedGoal": [],
      "impactedAssets": [],
      "statement": "A threat actor who is in a person-in-the-middle position between the user and the CloudFront Distribution can take advantage of a vulnerability in an older/depreciated SSL/TLS protocol, which leads to them being able manipulate view or modify the content served by the CloudFront Distribution"
    },
    {
      "id": "c7d756f0-0576-4644-a9b3-afea5bd52a7c",
      "numericId": 6,
      "displayOrder": 6,
      "tags": [
        "CloudFront"
      ],
      "threatSource": "threat actor",
      "prerequisites": "who is in a person-in-the-middle position between CloudFront Distribution and the origin",
      "threatAction": "take advantage of a vulnerability in an older/depreciated SSL/TLS protocol",
      "threatImpact": "them being able manipulate view or modify the content from the origin and served by the CloudFront Distribution",
      "impactedGoal": [],
      "impactedAssets": [],
      "statement": "A threat actor who is in a person-in-the-middle position between CloudFront Distribution and the origin can take advantage of a vulnerability in an older/depreciated SSL/TLS protocol, which leads to them being able manipulate view or modify the content from the origin and served by the CloudFront Distribution"
    },
    {
      "id": "d4378c9b-e680-41c1-8cc2-136e124461dd",
      "numericId": 7,
      "displayOrder": 7,
      "tags": [
        "CloudFront"
      ],
      "threatSource": "external threat actor",
      "threatAction": "bypass the CloudFront Distribution and associated controls (e.g. Geo-blocking)",
      "threatImpact": "direct access to static assets hosted on the Amazon S3 Origin",
      "impactedAssets": [
        "assets hosted on the S3 Origin for this application"
      ],
      "statement": "An external threat actor can bypass the CloudFront Distribution and associated controls (e.g. Geo-blocking), which leads to direct access to static assets hosted on the Amazon S3 Origin, negatively impacting assets hosted on the S3 Origin for this application"
    },
    {
      "id": "3baec34a-5ed0-459d-90d9-9e5497e29751",
      "numericId": 8,
      "displayOrder": 8,
      "tags": [
        "CloudFront"
      ],
      "threatSource": "external threat actor",
      "prerequisites": "with a network path to the CloudFront Distribution",
      "threatAction": "make arbitrary web requests that are either not explicitly allowed, blocked or counted",
      "threatImpact": "them perform reconnaissance and intrusion activities against the content served by the CloudFront Distribution",
      "impactedAssets": [
        "this application"
      ],
      "statement": "An external threat actor with a network path to the CloudFront Distribution can make arbitrary web requests that are either not explicitly allowed, blocked or counted, which leads to them perform reconnaissance and intrusion activities against the content served by the CloudFront Distribution, negatively impacting this application"
    },
    {
      "id": "7b1dbdd4-2cea-4b16-a15b-51b05822f972",
      "numericId": 9,
      "displayOrder": 9,
      "tags": [
        "CodeBuild"
      ],
      "threatSource": "threat actor",
      "prerequisites": "with physical access to CodeBuild storage",
      "threatAction": "view or modify build artefacts (e.g. cache, logs, exported raw test report data files, and build results)",
      "impactedGoal": [
        "confidentiality",
        "integrity"
      ],
      "impactedAssets": [
        "this application's build artefacts"
      ],
      "statement": "A threat actor with physical access to CodeBuild storage can view or modify build artefacts (e.g. cache, logs, exported raw test report data files, and build results), resulting in reduced confidentiality and/or integrity of this application's build artefacts"
    },
    {
      "id": "3e3ee2f1-053d-4e40-8917-52ce4f45ea56",
      "numericId": 10,
      "displayOrder": 10,
      "metadata": [
        {
          "key": "STRIDE",
          "value": [
            "T"
          ]
        }
      ],
      "tags": [
        "CodeBuild"
      ],
      "threatSource": "threat actor",
      "prerequisites": "who can trick a user into using a external vulnerable or malicious Docker image in their CodeBuild project",
      "threatAction": "tamper with the build process",
      "impactedAssets": [
        "this application"
      ],
      "statement": "A threat actor who can trick a user into using an external vulnerable or malicious Docker image in their CodeBuild project can tamper with the build process, negatively impacting this application"
    },
    {
      "id": "9d2c9e3b-f5c8-42af-8f63-4bd2309cf0d9",
      "numericId": 11,
      "displayOrder": 11,
      "metadata": [
        {
          "key": "STRIDE",
          "value": [
            "I",
            "E",
            "D"
          ]
        }
      ],
      "tags": [
        "CodeBuild"
      ],
      "threatSource": "threat actor",
      "prerequisites": "who is able to influence what commands are run within a CodeBuild job",
      "threatAction": "take run commands supported by 'privilegedMode'",
      "threatImpact": "them accessing the Docker API and associated resources",
      "impactedAssets": [
        "this application"
      ],
      "statement": "A threat actor who is able to influence what commands are run within a CodeBuild job can take run commands supported by 'privilegedMode', which leads to them accessing the Docker API and associated resources, negatively impacting this application"
    },
    {
      "id": "9893d5cc-6cff-4d3e-8bd2-f7c2f4af58f2",
      "numericId": 12,
      "displayOrder": 12,
      "tags": [
        "Cognito"
      ],
      "threatSource": "external threat actor",
      "prerequisites": "with a network path to the API",
      "threatAction": "access the API using a method other than Cognito user pool authentication",
      "threatImpact": "the actor being able to do anything a authenticated or unauthenticated user is permitted to do",
      "impactedAssets": [
        "this application"
      ],
      "statement": "An external threat actor with a network path to the API can access the API using a method other than Cognito user pool authentication, which leads to the actor being able to do anything an authenticated or unauthenticated user is permitted to do, negatively impacting this application"
    },
    {
      "id": "9bbb493a-c435-420a-aa35-3fa992064127",
      "numericId": 13,
      "displayOrder": 13,
      "tags": [
        "EBS",
        "SSE",
        "EC2"
      ],
      "threatSource": "threat actor",
      "prerequisites": "with physical access to EBS (Elastic Block Store) storage",
      "threatAction": "view or modify the contents on the disk",
      "impactedGoal": [
        "confidentiality",
        "integrity"
      ],
      "impactedAssets": [
        "data within the EBS volume"
      ],
      "statement": "A threat actor with physical access to EBS (Elastic Block Store) storage can view or modify the contents on the disk, resulting in reduced confidentiality and/or integrity of data within the EBS volume"
    },
    {
      "id": "8c3352e2-b817-4139-9256-81e75ff127f7",
      "numericId": 14,
      "displayOrder": 14,
      "tags": [
        "EC2",
        "VPC"
      ],
      "threatSource": "external threat actor",
      "threatAction": "discover EC2 instances that have public IP addresses and allow ingress to all internet addresses, or move laterally to non-public EC2 instances",
      "threatImpact": "reconnaissance and intrusion activities being performed against the exposed attack surface",
      "impactedAssets": [
        "this application"
      ],
      "statement": "An external threat actor can discover EC2 instances that have public IP addresses and allow ingress to all internet addresses, or move laterally to non-public EC2 instances, which leads to reconnaissance and intrusion activities being performed against the exposed attack surface, negatively impacting this application"
    },
    {
      "id": "310b7a7f-e13c-44a3-bb3a-91ad7bf9ba98",
      "numericId": 15,
      "displayOrder": 15,
      "tags": [
        "EC2",
        "VPC"
      ],
      "threatSource": "threat actor",
      "prerequisites": "with established persistence within a host in the VPC",
      "threatAction": "go unnoticed for a prolonged period due unnecessary Security Groups being kept as their rationale for existence is not clear",
      "statement": "A threat actor with established persistence within a host in the VPC can go unnoticed for a prolonged period due unnecessary Security Groups being kept as their rationale for existence is not clear"
    },
    {
      "id": "1c62b52e-f3ba-4968-a14d-745691b706f6",
      "numericId": 16,
      "displayOrder": 16,
      "tags": [
        "IAM"
      ],
      "threatSource": "threat actor",
      "prerequisites": "with access to a valid IAM Principal",
      "threatAction": "take advantage of unnecessary permissions included in the associated IAM Policy",
      "threatImpact": "the actor being able to do anything the user can do",
      "impactedAssets": [
        "this application"
      ],
      "statement": "A threat actor with access to a valid IAM Principal can take advantage of unnecessary permissions included in the associated IAM Policy, which leads to the actor being able to do anything the user can do, negatively impacting this application"
    },
    {
      "id": "5d521ccc-ca9c-4790-b045-0a1e26de9aa0",
      "numericId": 17,
      "displayOrder": 17,
      "tags": [
        "Lambda",
        "Vuln Management"
      ],
      "threatSource": "threat actor",
      "prerequisites": "with the ability to invoke or interact with a Lambda function",
      "threatAction": "take advantage of a disclosed vulnerability within the runtime version used for a non-container based Lambda function",
      "impactedGoal": [
        "confidentiality",
        "integrity",
        "availability"
      ],
      "impactedAssets": [
        "this application"
      ],
      "statement": "A threat actor with the ability to invoke or interact with a Lambda function can take advantage of a disclosed vulnerability within the runtime version used for a non-container based Lambda function, resulting in reduced confidentiality, integrity and/or availability of this application"
    },
    {
      "id": "ef634b32-084b-42f6-ae5c-4e65b3953ba5",
      "numericId": 18,
      "displayOrder": 18,
      "tags": [
        "RDS",
        "Network Access"
      ],
      "threatSource": "threat actor",
      "prerequisites": "with a network path to the RDS cluster listener",
      "threatAction": "perform a network probe the database using it's default port",
      "threatImpact": "perform reconnaissance and intrusion activities against the exposed attack surface",
      "impactedAssets": [
        "this application"
      ],
      "statement": "A threat actor with a network path to the RDS cluster listener can perform a network probe the database using it's default port, which leads to perform reconnaissance and intrusion activities against the exposed attack surface, negatively impacting this application"
    },
    {
      "id": "172ce9d2-8159-4c70-a2b7-70dc846bdf70",
      "numericId": 19,
      "displayOrder": 19,
      "tags": [
        "RDS",
        "SSE"
      ],
      "threatSource": "threat actor",
      "prerequisites": "with access to underlying storage used for the RDS cluster",
      "threatAction": "view the contents of the database",
      "impactedGoal": [
        "confidentiality"
      ],
      "impactedAssets": [
        "this application's data"
      ],
      "statement": "A threat actor with access to underlying storage used for the RDS cluster can view the contents of the database, resulting in reduced confidentiality of this application's data"
    },
    {
      "id": "8c383856-988d-45a4-bc92-5c9aab42aab3",
      "numericId": 20,
      "displayOrder": 20,
      "metadata": [
        {
          "key": "STRIDE",
          "value": [
            "T",
            "I"
          ]
        }
      ],
      "tags": [
        "S3",
        "Public Access"
      ],
      "threatSource": "threat actor",
      "prerequisites": "who has discovered a S3 bucket configured for public read or write",
      "threatAction": "read or write data to or from the S3 bucket",
      "impactedGoal": [
        "confidentiality",
        "integrity",
        "availability"
      ],
      "impactedAssets": [
        "this application's data"
      ],
      "statement": "A threat actor who has discovered a S3 bucket configured for public read or write can read or write data to or from the S3 bucket, resulting in reduced confidentiality, integrity and/or availability of this application's data"
    },
    {
      "id": "1b240cb7-3557-41e7-9c19-21ebc73ad0ec",
      "numericId": 21,
      "displayOrder": 21,
      "metadata": [
        {
          "key": "STRIDE",
          "value": [
            "R"
          ]
        }
      ],
      "tags": [
        "S3",
        "Repudiation"
      ],
      "threatSource": "threat actor",
      "prerequisites": "who is authenticated and authorized",
      "threatAction": "deny taking a action associated with an Amazon S3 bucket",
      "threatImpact": "lack of knowledge/proof of who took the action",
      "impactedGoal": [],
      "impactedAssets": [],
      "statement": "A threat actor who is authenticated and authorized can deny taking an action associated with an Amazon S3 bucket, which leads to lack of knowledge/proof of who took the action"
    },
    {
      "id": "77e519f0-641e-4b99-96a3-db0130381c97",
      "numericId": 22,
      "displayOrder": 22,
      "metadata": [
        {
          "key": "STRIDE",
          "value": [
            "I",
            "T"
          ]
        }
      ],
      "tags": [
        "S3",
        "MiTM"
      ],
      "threatSource": "threat actor",
      "prerequisites": "who is in a person-in-the-middle position between the client and the Amazon S3 bucket endpoint",
      "threatAction": "view plaintext requests and responses",
      "threatImpact": "them being able manipulate view or modify the requests or responses",
      "impactedGoal": [],
      "impactedAssets": [
        "this application's data"
      ],
      "statement": "A threat actor who is in a person-in-the-middle position between the client and the Amazon S3 bucket endpoint can view plaintext requests and responses, which leads to them being able manipulate view or modify the requests or responses, negatively impacting this application's data"
    },
    {
      "id": "cc8b9fcb-dd2a-41ad-a4f1-11e93ae0b031",
      "numericId": 23,
      "displayOrder": 23,
      "metadata": [
        {
          "key": "STRIDE",
          "value": [
            "I"
          ]
        }
      ],
      "tags": [
        "S3",
        "Network Access"
      ],
      "threatSource": "external threat actor",
      "prerequisites": "who has discovered misconfigured S3 hosted website",
      "threatAction": "bypass the protections put in place by CloudFront (e.g. geo blocking)",
      "impactedGoal": [
        "confidentiality"
      ],
      "impactedAssets": [
        "this application's data"
      ],
      "statement": "An external threat actor who has discovered misconfigured S3 hosted website can bypass the protections put in place by CloudFront (e.g. geo blocking), resulting in reduced confidentiality of this application's data"
    },
    {
      "id": "90619d5b-6450-4108-8013-2eaafc5788b5",
      "numericId": 24,
      "displayOrder": 24,
      "metadata": [
        {
          "key": "STRIDE",
          "value": [
            "S",
            "T",
            "I",
            "E"
          ]
        }
      ],
      "tags": [
        "Secrets Manager",
        "AuthN"
      ],
      "threatSource": "threat actor",
      "prerequisites": "who has obtained a long-lived secret (e.g. password)",
      "threatAction": "spoof an authenticated user or system",
      "threatImpact": "the actor being able to do anything the user or system can do",
      "impactedAssets": [
        "this application"
      ],
      "statement": "A threat actor who has obtained a long-lived secret (e.g. password) can spoof an authenticated user or system, which leads to the actor being able to do anything the user or system can do, negatively impacting this application"
    }
  ]
}"
`;
